### PR TITLE
[TailCallElim] Don’t mark llvm.stackrestore with tail-call 

### PR DESCRIPTION
**Pull request by scui-ibm**
This is to teach tailcallelim transformation not to mark llvm.stackrestore with tail-call, as the intrinsic call can modify unescaped allocas from the caller.
 
This fixes a problem found with our downstream testing. The problem can also be shown when running the test case llvm/test/Transforms/MemCpyOpt/stackrestore with passes=”tailcallelim, memcpyopt”.

### CODE REVIEW

**Review by efriedma-quic — COMMENTED** [review]
The alternative here is that we declare the tailcallelim transform on llvm.stackrestore illegal.  Which I think is a better idea in case anything else looks at the "tail" marker.

**Review by nikic — COMMENTED** [review]
> The alternative here is that we declare the tailcallelim transform on llvm.stackrestore illegal. Which I think is a better idea in case anything else looks at the "tail" marker.

Agree. LangRef says this:

> Both markers imply that the callee does not access allocas from the caller.

Where "both markers" are tail/musttail here. This requirement is not satisfied for stackrestore, so it should not get a tail marker.

**Comment by scui-ibm** [comment]
Thanks to @efriedma-quic and @nikic . I'll make a change in tailcallelim instead for this

**Comment by nikic** [comment]
Can you please update the PR title and description?

File: llvm/lib/Transforms/Scalar/TailRecursionElimination.cpp [hunk 1]
```diff
@@ -243,6 +243,11 @@ static bool markTails(Function &F, OptimizationRemarkEmitter *ORE) {
           isa<PseudoProbeInst>(&I))
         continue;
 
+      // Bail out intrinsic stackrestore call.
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/TailRecursionElimination.cpp:246]: I'd add "because it can modify unescaped allocas" or so here.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**